pkg/kubelet/cm/memorymanager/policy_static.go

```diff
@@ -788,3 +857,120 @@ func (p *staticPolicy) GetAllocatableMemory(s state.State) []state.Block {
 	}
 	return allocatableMemory
 }
+
+func (p *staticPolicy) updateStateToReuse(s state.State, pod *v1.Pod, container *v1.Container, memoryBlocks []state.Block) {
+	podUID := string(pod.UID)
+
+	// If pod entries to p.reusableState other than the current pod exist, delete them
+	for uid := range p.reusableState {
+		if podUID != uid {
+			delete(p.reusableState, podUID)
+		}
+	}
+
+	if isInitContainer(pod, container) {
+		// update the reusable state to the state before the init containers allocation
+		if _, ok := p.reusableState[podUID]; !ok {
+			p.reusableState[podUID] = s.GetMachineState()
+		}
+		return
+	}
+
+	if _, ok := p.reusableState[podUID]; !ok {
+		return
+	}
+
+	// Otherwise it is an app container.
+	// Update the re-usable machine state
+	for _, memoryBlock := range memoryBlocks {
+		p.updateMachineState(p.reusableState[podUID], memoryBlock.NUMAAffinity, memoryBlock.Type, memoryBlock.Size)
+	}
+}
+
+func (p *staticPolicy) updateInitContainersReusableMemory(pod *v1.Pod, container *v1.Container, memoryBlocks []state.Block) {
```

Comment:
yes exactly!

Reply:
I'm wondering if there is a better way to abstract this. Although I get what it's doing, it's not intuitive that these two things should happen together, nor is it obvious from the name of the function that this is what is going on.

As mentioned in another comment. I wonder if part (2) of this can be folded into `getPodReusableMemory()` and then this function just renamed / reduced to:
```
func (p *staticPolicy) addInitContainersReusableMemory(pod *v1.Pod, container *v1.Container, memoryBlocks []state.Block) {

	podUID := string(pod.UID)

	if isInitContainer(pod, container) {
		if _, ok := p.initContainersReusableMemory[podUID]; !ok {
			p.initContainersReusableMemory[podUID] = []map[string][]state.Block{}
		}

		p.initContainersReusableMemory[podUID] = append(p.initContainersReusableMemory[podUID], map[string][]state.Block{
			container.Name: memoryBlocks,
		})
	}
}
```

Reply:
see my comment above, we can not join fetch and update under the same method because it called from two different methods `GetTopologyHints`(should not update anything) and `Allocate`
btw CPU manager has similar method https://github.com/kubernetes/kubernetes/blob/756203fda0f22a093b4f71022e2aa6d12743e840/pkg/kubelet/cm/cpumanager/policy_static.go#L194 just calculation is much simpler :(

Reply:
Hey. Looking again at `updateInitContainersReusableMemory()`. Sorry, still trying to wrap my head around why it needs to do anything different than what the CPUManager’s updateCPUsToReuse() does (except for with memory instead of CPUs).

In the CPUManager’s implementation, we:
1. Track the total reusable CPUs at the pod level (not per container)
2. Unconditionally union any CPUs allocated to init containers into the reusable set
3. Unconditionally subtract out and CPUs allocated to app containers

In the memory manager, you track the reusable memory blocks allocated to each init container separately, which I think unnecessarily complicates the logic (and has likely been the source of alot of my confusion). Is there any reason you can’t follow some similar logic for memory that tracks "total" reusable memory instead of per init container? That would turn the `updateInitContainersReusableMemory()` function into something like:
```
func (p *staticPolicy) updateInitContainersReusableMemory(pod *v1.Pod, container *v1.Container, memoryBlocks []state.Block) {
	// If pod entries to p.initContainersReusableMemory other than the current pod exist, delete them.
	for podUID := range p.initContainersReusableMemory {
		if podUID != string(pod.UID) {
			delete(p.initContainersReusableMemory, podUID)
		}
	}
	// If no memory blocks map exists in initContainersReusableMemory by this pod yet, create one.
	if _, ok := p.initContainersReusableMemory[string(pod.UID)]; !ok {
		p.initContainersReusableMemory[string(pod.UID)] = []map[string][]state.Block{}
	}
	// Check if the container is an init container.
	// If so, "merge" it with the set of memory blocks already passed in by other init containers
	// i.e. keep the larger of the two allocations for each matching block
	if isInitContainer(pod, container) {
		// Walk through `memoryBlocks` and `p.initContainersReusableMemory[string(pod.UID)]`
		// If the blocks match by type and NUMA affinity
		//     If memoryBlocks.size > initContainersReusableMemory[string(pod.UID)].size
		//         initContainersReusableMemory[string(pod.UID)].size = memoryBlocks.size
	}
	// Otherwise it is an app container.
	// Remove memory from the matching blocks of reusable memory for any new allocations.
		// Walk through `memoryBlocks` and `p.initContainersReusableMemory[string(pod.UID)]`
		// If the blocks match by type and NUMA affinity
		//     if memoryBlocks.size >= p.initContainersReusableMemory[string(pod.UID)].size {
		//         p.initContainersReusableMemory[string(pod.UID)].size = 0
		//     } else {
		//         p.initContainersReusableMemory[string(pod.UID)].size -= memoryBlocks.size
		//     }
}
```
To me this feels much more intuitive as we keep a running tally of how much memory is being consumed for each matching block type, rather than trying to (artifically) consolidate things across the init containers they were originally allocated to.

Reply:
To reuse the logic of detecting matching blocks you could write a helper like:
```
func walkMatchingMemoryBlocks(m1, m2 map[string][]state.Block, f func(b1, b2 state.Block)) {
	for _, b1s := range m1 {
		for b2s:= range m2 {
			for i, b1 := range b1s {
				for j, b2 := range b2s {
					if b1.Type != b2.Type {
						continue
					}
					if !isNUMAAffinitiesEqual(b1.NUMAAffinity, b2.NUMAAffinity) {
						continue
					}
					f(b1s[i], b2s[j])
				}
			}
		}
	}
}
```
And then use it in the logic in the previous comment as:
```
walkMatchingMemoryBlocks(memoryBlocks, initContainersReusableMemory[string(pod.UID)], func(b1, b2 state.Block)) {
	...
})
```

Reply:
But under your implementation, you do not really know from which NUMA how many memory you can reuse. Because in general when we are talking about the container scope each init container can start on the different NUMA node(depends on the memory fragmentation). Also take into account that you have different memory types(memory, hugepages-2Mi, hugepages-1Gi).
IIRC under the CPU manager, the hint calculation will reject any hints that do not re-use the memory so you will end with the admission error - https://github.com/kubernetes/kubernetes/issues/94220.

Reply:
> you do not really know from which NUMA how many memory you can reuse. Because in general when we are talking about the container scope each init container can start on the different NUMA node(depends on the memory fragmentation).

That should not be a concern of this function -- the memory has already been allocated to the container before this function is ever called (i.e. by `getPodReusableMemory()`) -- this function just cleans up the state in `initContainersReusableMemory` to the set of reusable memory from any new allocations that have just been made.

> Also take into account that you have different memory types(memory, hugepages-2Mi, hugepages-1Gi).

As written, this should account for that. That is what is meant by the comment about:
```
If the blocks match by type and NUMA affinity
...
```

Reply:
Maybe this is where the confusion is, but I'm expecting the following:
1. `initContainersReusableMemory` contains the current state of reusable memory that a container can reusse when making allocation decisions
2. `getPodReusableMemory()` uses `initContainersReusableMemory` (plus other logic) to decide exaclty what memory to allocate to a container
3. Later on `updateInitContainersReusableMemory()` is called with the *actual* set of memory blocks allocated to the container so that the state of `initContainersReusableMemory()` can be updated appropriately.

Reply:
I currently trying to update the code according to the comment, and one problem that I encounter, I still need to update the init container memory assignments once I start new init or app container.

Like for the CPU manager, something like `{init1: 4-7, app1: 4-7}` makes total sense and reflects correct the machine state.
But for the memory manager `{init1: 1G, app1: 1G}` does not reflect the real machine state, so during the allocation of the container app1 we need to update init1 memory to 0(memory re-used).

Reply:
Why does that not reflect the real machine state? They are both allocated that much memory at the the time they each run.

Is it because of the way memory accounting is done internally? I.e. there is no way to „discretely“ divide up the memory (it is just a count/size), so there is no way to know that that memory is the „same“ memory (rather than being additive).

Reply:
If so, that feels like a Limitation of the data structure used to represent the memory allocations (I.e. it wasn’t designed to account for memory reuse across containers).

Althogh the accounting would be „correct“ in terms of total memory allocated if you set the init containers allocation to 0, I find it equally confusing to see that set to 0 the b the final machine state even though it clearly had/has some memory allocated to it.

Reply:
> Is it because of the way memory accounting is done internally? I.e. there is no way to „discretely“ divide up the memory (it is just a count/size), so there is no way to know that that memory is the „same“ memory (rather than being additive).

Yes, it exactly the reason, the allocation with 0 value will stay under the state file only until the next container started.
I can avoid setting it to 0 and it will make logic much simplier, but in this case, we will have different numbers for `sum(requests_accross_all_containers)` and machine free memory.